### PR TITLE
Expand unit test coverage

### PR DESCRIPTION
**Pull request by jmjatlanta**
While not strictly "unit testing", more like mini system/integration testing, developers sometimes need to test their code alongside portions of the existing system. In the case that brought about this change, a new CryptoCondition needed to be tested in the context of the whole create transaction/spend transaction process.

testutils already existed to do some of this, but it did not seem to be used often. Existing functions were collected and adjusted to work inside a class. Now simple chain transactions can be tested by writing a gtest:
```
TestChain chain; // start a new chain

// Basic user wallets
auto notary = chain.AddWallet(chain.getNotaryKey());
auto alice = chain.AddWallet();
auto bob = chain.AddWallet();

CBlock lastBlock = chain.generateBlock(); // Genesis block

// turn on CryptoConditions
ASSETCHAINS_CC = 1;

{
    // create a transaction that gives some coin to Alice
    CValidationState returnState = notary->Transfer(alice, 100000);
    // verify everything worked
    EXPECT_TRUE( returnState.IsValid() );
    // mine a block
    lastBlock = chain.generateBlock(); // vtx[0] is unrelated, vtx[1] is ours
}
{
    // notary gives some coin to Bob, with the stipulation that he pay some kind of miner fee
    auto notaryPrevOut = notary->GetAvailable(200000);
    CMutableTransaction tx;
    CTxIn notaryIn;
    notaryIn.prevout.hash = notaryPrevOut.first.GetHash();
    notaryIn.prevout.n = notaryPrevOut.second;
    tx.vin.push_back(notaryIn);
    // give some coin to bob, but he must pay something to the miners
    CC* bobMustPayFee = MakeCCcond1(EVAL_INCLMINERFEE, bob->GetPubKey());
    CTxOut bobOut;
    bobOut.scriptPubKey = CCPubKey(bobMustPayFee);
    bobOut.nValue = 200000;
    tx.vout.push_back(bobOut);
    // give the rest back to the notary
    CTxOut notaryOut;
    notaryOut.scriptPubKey = GetScriptForDestination(notary->GetPubKey());
    notaryOut.nValue = notaryPrevOut.first.vout[1].nValue - 200000;
    tx.vout.push_back(notaryOut);
    // sign it
    uint256 hash = SignatureHash(notaryPrevOut.first.vout[1].scriptPubKey, tx, 0, SIGHASH_ALL, 0, 0);
    tx.vin[0].scriptSig << notary->Sign(hash, SIGHASH_ALL);
    // send it to the mempool
    CTransaction fundBob(tx);
    CValidationState returnState = chain.acceptTx(fundBob);
    // verify everything worked
    EXPECT_TRUE( returnState.IsValid() );
    // mine a block
    CBlock currentBlock = chain.generateBlock();
    // the wallet doesn't understand these transactions, so add this vout to Bob's wallet manually
    bob->AddOut(currentBlock.vtx[1], 0);
    cc_free(bobMustPayFee);
}
```

### CODE REVIEW

**Comment by jmjatlanta** [comment]
I believe this PR to be complete, and ready for review.

You may question why move bitcoin and zcash tests into the test-komodo directory instead of their own directory. Komodo has changed significantly enough that getting all tests to work in those two areas would be difficult. I see two options:
1. Disable individual tests that simply do not work, run bitcoin, zcash and komodo tests as 3 executables.
2. Never run bitcoin or zcash tests, maintain 1 test suite that runs our tests.

Both have pros and cons. I opted for 2. I still debate which is best.

**Comment by dimxy** [comment]
this should be rebased to include the S6 updates

File: src/komodo_globals.h [hunk 1]
```diff
@@ -42,18 +42,8 @@ int32_t NUM_PRICES; uint32_t *PVALS;
...
-const uint32_t nS4Timestamp = 1592146800; //dPoW Season 4 2020 hardfork Sunday, June 14th, 2020 03:00:00 PM UTC
-const int32_t nS4HardforkHeight = 1922000;   //dPoW Season 4 2020 hardfork Sunday, June 14th, 2020 
-
-const uint32_t nS5Timestamp = 1623682800;  //dPoW Season 5 Monday, June 14th, 2021 (03:00:00 PM UTC)
-const int32_t nS5HardforkHeight = 2437300;  //dPoW Season 5 Monday, June 14th, 2021
-
-#define _COINBASE_MATURITY 100
-int COINBASE_MATURITY = _COINBASE_MATURITY;//100;
-unsigned int WITNESS_CACHE_SIZE = _COINBASE_MATURITY+10;
+unsigned int WITNESS_CACHE_SIZE = 100+10; // coinbase maturity plus 10
```
Comment (dimxy) [inline src/komodo_globals.h:46]: a named constant instead of '100'?

File: src/Makefile.gtest.include [hunk 2]
```diff
@@ -2,47 +2,43 @@
...
 # file COPYING or https://www.opensource.org/licenses/mit-license.php .
 
-TESTS += komodo-gtest
-bin_PROGRAMS += komodo-gtest
+TESTS += zcash-gtest
+bin_PROGRAMS += zcash-gtest
 
 # tool for generating our public parameters
-komodo_gtest_SOURCES = \
+zcash_gtest_SOURCES = \
```
Comment (dimxy) [inline src/Makefile.gtest.include:9]: aha, renamed to zcash_gtest as in fact it tests just zcash features

File: src/komodo_globals.h (hunk 1 above)
Comment (jmjatlanta) [reply src/komodo_globals.h]: I would like to replace it with something other than a global. It is only used in wallet functions, so I put it there. See https://github.com/KomodoPlatform/komodo/pull/453/commits/8db798bf3355f791134f7e32e52f00950bb7ee40

File: src/main.h [hunk 3]
```diff
@@ -78,7 +77,7 @@ static const bool DEFAULT_ALERTS = true;
 /** Minimum alert priority for enabling safe mode. */
 static const int ALERT_PRIORITY_SAFE_MODE = 4000;
 /** Maximum reorg length we will accept before we shut down and alert the user. */
-static unsigned int MAX_REORG_LENGTH = _COINBASE_MATURITY - 1;
+static unsigned int MAX_REORG_LENGTH = 100 - 1; // based on COINBASE_MATURITY
```
[the remark this answers is not in the document]
Comment (jmjatlanta) [reply src/main.h:80]: Removed MAX_REORG_LENGTH, see comment above

**Comment by tonymorony** [comment]
these changes implemented in combined PR: https://github.com/KomodoPlatform/komodo/pull/559

**Pull request closed without merging**